src/predictor/gpu_predictor.cu

```diff
@@ -237,6 +237,19 @@ class GPUPredictor : public xgboost::Predictor {
     }
   }
 
+  void PredictionDeviceOffsets(size_t total_size, size_t batch_offset, size_t batch_size,
+                               int n_classes, std::vector<size_t>* out_offsets) {
+    auto& offsets = *out_offsets;
+    size_t n_shards = devices_.Size();
+    offsets.resize(n_shards + 2);
+    size_t rows_per_shard = (batch_size + n_shards - 1) / n_shards;
+    for (int shard = 0; shard < devices_.Size(); ++shard) {
+      offsets[shard] = batch_offset + shard * rows_per_shard * n_classes;
+    }
+    offsets[n_shards] = batch_offset + batch_size * n_classes;
```

Comment:
if the batch_size is fairly close to the number of gpu's (say, slightly more), then `offsets[n_shards]` can be less than the one before it. this is possible for the last batch in the dmatrix

Reply:
Now limit it to `batch_size`.